core/src/main/java/org/infinispan/functional/impl/MetaParams.java

```diff
@@ -50,94 +48,186 @@
  * instances, they cannot act on the globally at the {@link MetaParams} level,
  * and hence there is no risk of users misusing {@link MetaParams}.
  *
+ * This class should not be accessible from user code, therefore it is package-protected.
+ *
  * @since 8.0
  */
 @NotThreadSafe
 @Experimental
-public final class MetaParams {
+final class MetaParams implements Iterable<MetaParam<?>> {
 
    private static final MetaParam<?>[] EMPTY_ARRAY = {};
    private MetaParam<?>[] metas;
+   private int length;
 
-   private MetaParams(MetaParam<?>[] metas) {
+   private MetaParams(MetaParam<?>[] metas, int length) {
       this.metas = metas;
+      this.length = length;
+      assert checkLength();
    }
 
    public boolean isEmpty() {
-      return metas.length == 0;
+      return length == 0;
    }
 
    public int size() {
-      return metas.length;
+      return length;
    }
 
    public MetaParams copy() {
-      return new MetaParams(Arrays.copyOf(metas, metas.length));
+      if (length == 0) {
+         return empty();
+      } else {
+         return new MetaParams(Arrays.copyOf(metas, metas.length), length);
+      }
    }
 
-   public <T> Optional<T> find(Class<T> type) {
+   public <T extends MetaParam> Optional<T> find(Class<T> type) {
       return Optional.ofNullable(findNullable(type));
    }
 
    @SuppressWarnings("unchecked")
-   private <T> T findNullable(Class<T> type) {
+   private <T extends MetaParam> T findNullable(Class<T> type) {
       for (MetaParam<?> meta : metas) {
-         if (meta.getClass().isAssignableFrom(type))
+         if (meta != null && meta.getClass().isAssignableFrom(type))
             return (T) meta;
       }
 
       return null;
    }
 
-   public void add(MetaParam.Writable meta) {
```

Comment:
`MetaParamsInternalMetadata.Builder` implements `Metadata.Builder`, and one method in that (not published in this PR) calls `MetaParams.replace` when it's building the metadata object. The type I use there shouldn't be settable by the user (it's internal info) so it does not implement `MetaParam.Writable`.
I need only replace, but `add`, `addMany` and `replace` should use the same type. And I've changed `MetaParams` to be non-public class, so all modifications should be done only through checked interfaces (using `Writable`).